src/dataflow/src/render/reduce.rs

```diff
@@ -87,20 +87,39 @@ where
             // Our first step is to extract `(key, vals)` from `input`.
             // We do this carefully, attempting to avoid unneccesary allocations
             // that would result from cloning rows in input arrangements.
-            let group_key_clone = group_key.clone();
-            let aggregates_clone = aggregates.clone();
-
-            // Tracks the required number of columns to extract.
-            let mut columns_needed = 0;
-            for key in group_key.iter() {
-                for column in key.support() {
-                    columns_needed = std::cmp::max(columns_needed, column + 1);
-                }
+
+            let input_arity = input.arity();
+
+            // Form an operator for evaluating key expressions.
+            let mut key_mfp = expr::MapFilterProject::new(input_arity)
+                .map(group_key.iter().cloned())
+                .project(input_arity..input_arity + group_key.len());
+
+            // Form an operator for evaluating value expressions.
+            let mut val_mfp = expr::MapFilterProject::new(input_arity)
+                .map(aggregates.iter().map(|a| a.expr.clone()))
+                .project(input_arity..input_arity + aggregates.len());
+
+            // Determine the columns we'll need from the row.
+            let mut demand = Vec::new();
+            demand.extend(key_mfp.demand());
+            demand.extend(val_mfp.demand());
+            demand.sort();
+            demand.dedup();
+            // remap column references to the subset we use.
+            let mut demand_map = std::collections::HashMap::new();
+            for column in demand.iter() {
+                demand_map.insert(*column, demand_map.len());
             }
-            for aggr in aggregates.iter() {
-                for column in aggr.expr.support() {
-                    columns_needed = std::cmp::max(columns_needed, column + 1);
-                }
+            key_mfp.permute(&demand_map, demand_map.len());
+            key_mfp.optimize();
+            val_mfp.permute(&demand_map, demand_map.len());
+            val_mfp.optimize();
+
+            // transform `demand` into "skips" we'll do on an iterator.
+            for index in (1..demand.len()).rev() {
```

Comment:
so just to make sure I understand right the intent here would be to take a demand like

```
[ 2, 5, 7, 10, 13, 14]
```
and turn it into

```
[2, 2, 1, 2, 2, 0]
```

imo it would be a little bit nicer if the skips were a separate array - afaict this is only an issue in the planning step and not on the request path?

Reply:
I put in a 
```rust
            // rename as skips to reflect their new role.
            let skips = demand;
```
but I can deconstruct and reconstruct the allocation if that is even more clear.

Reply:
that works for me